### PR TITLE
Gray Binary Converter

### PR DESCRIPTION
**Pull request by quekyj**
## Description & Motivation

Add converter for binary-to-gray and gray-to-binary.

## Related Issue(s)

#50 

## Testing

Test is compared with manually hand calculated data.

## Backwards-compatibility

> Is this a breaking change that will not be backwards-compatible? If yes, how so?

No

## Documentation

> Does the change require any updates to documentation? If so, where? Are they included?

Yes.

### CODE REVIEW

File: lib/src/binary_gray.dart [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  ///
+  /// For each bit in the [binary] input, starting from the least significant
+  /// bit (index 0), the function calculates the corresponding Gray code bit
+  /// based on XOR operation with the previous bit. The resulting Gray code
+  /// bits are returned in a list.
+  ///
+  /// If [binary] has no bits, an empty list is returned.
+  ///
+  /// Returns a list of [LogicValue] representing the Gray code.
+  List<LogicValue> binaryToGrayMap(Logic binary) {
```
Comment (mkorbel1) [inline lib/src/binary_gray.dart:63]: This should take `LogicValue binary`

File: test/binary_gray_test.dart [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+//
+// 2023 October 5
+// Author: Yao Jing Quek <yao.jing.quek@intel.com>
+//
+
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/rohd_hcl.dart';
+import 'package:test/test.dart';
+
+Future<void> main() async {
```
Comment (mkorbel1) [inline test/binary_gray_test.dart:15]: An interesting test might be to ensure that no sequential encodings have more than 1 bit difference and iterate through some?  Just to test the intent of the grey code regardless of the algorithm?

**Review by mkorbel1 — CHANGES_REQUESTED** [review]

File: lib/src/binary_gray.dart [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+  ///
+  /// For each bit in the [binary] input, starting from the least significant
+  /// bit (index 0), the function calculates the corresponding Gray code bit
+  /// based on XOR operation with the previous bit. The resulting Gray code
+  /// bits are returned in a list.
+  ///
+  /// If [binary] has no bits, an empty list is returned.
+  ///
+  /// Returns a list of [LogicValue] representing the Gray code.
+  List<LogicValue> binaryToGrayMap(LogicValue binary) {
```
Comment (mkorbel1) [inline lib/src/binary_gray.dart:68]: could this be made into a `static` function?  And we can call it just `binaryToGray`?  Seems like a useful function to expose.

Comment (mkorbel1) [inline lib/src/binary_gray.dart:68]: This should just return one `LogicValue`, do the swizzle inside the function.

File: lib/src/binary_gray.dart [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  ///
+  /// For each bit in the [gray] input, starting from the least significant bit
+  /// (index 0), the function calculates the corresponding binary bit based
+  /// on XOR operation with the previous binary bit. The resulting binary bits
+  /// are returned in a list.
+  ///
+  /// If [gray] has no bits, an empty list is returned.
+  ///
+  /// Returns a list of [LogicValue] representing the binary representation.
+  List<LogicValue> grayToBinaryMap(LogicValue gray) {
```
Comment (mkorbel1) [inline lib/src/binary_gray.dart:139]: same as above, remove the word `Map`, make it `static`, return `LogicValue`

File: lib/src/binary_gray.dart [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+  /// Creates a [BinaryToGrayConverter] instance with the specified binary
+  /// input.
+  ///
+  /// The [binary] parameter is the binary input that you want to convert
+  /// to Gray code. The width of the input [binary] determines the width
+  /// of the Gray code output.
+  BinaryToGrayConverter(Logic binary) {
+    final inputWidth = binary.width;
+    binary = addInput('binary_input', binary, width: inputWidth);
```
Comment (mkorbel1) [inline lib/src/binary_gray.dart:35]: for consistency, let's make these ports `'binary'` and `'gray'`, like they are in the GrayToBinary?

File: lib/src/binary_gray.dart [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+/// code value as input and produces the equivalent binary representation as
+/// output. It internally uses the [grayToBinaryMap] function for the
+/// conversion
+class GrayToBinaryConverter extends Module {
+  /// The binary representation output of the converter.
+  ///
+  /// This [Logic] value represents the binary representation of the Gray code
+  /// input provided to the converter. It can be accessed using the [binaryVal]
+  /// getter.
+  Logic get binaryVal => output('binary');
```
Comment (mkorbel1) [inline lib/src/binary_gray.dart:96]: Why not just `binary` instead of `binaryVal`?

File: lib/src/binary_gray.dart [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+/// value as input and produces the equivalent Gray code representation as
+/// output. It internally uses the [binaryToGrayMap] function for the
+/// conversion.
+class BinaryToGrayConverter extends Module {
+  /// The Gray code representation output of the converter.
+  ///
+  /// This [Logic] value represents the Gray code representation of the binary
+  /// input provided to the converter. It can be accessed using
+  /// the [grayCode] getter.
+  Logic get grayCode => output('gray_code');
```
Comment (mkorbel1) [inline lib/src/binary_gray.dart:25]: How about `gray` instead of `grayCode`?

**Review by mkorbel1 — CHANGES_REQUESTED** [review]
Looks really good, couple small adjustments

**Comment by quekyj** [comment]
@mkorbel1 any idea why this test fail?

```cmd
00:09 +120 -1: test\configurator_test.dart: fifo configurator should generate FIFO [E]
  Expected: contains '(wrPointer == 3\'h6'
    Actual: '/**\n'
              ' * Generated by ROHD - www.github.com/intel/rohd\n'
              ' * Generation time: 2024-01-12 10:14:52.310 +08:00\n'
              ' * ROHD Version: 0.5.0\n'
              ' */\n'
              '\n'
              'module RegisterFile(\n'
              'input logic [5:0] wr_data_0,\n'
              'input logic [2:0] rd_addr_0,\n'
              'input logic [2:0] wr_addr_0,\n'
              'input logic clk,\n'
              'input logic reset,\n'
              'input logic wr_en_0,\n'
              'input logic rd_en_0,\n'
              'output logic [5:0] rd_data_0\n'
              ');\n'
              'logic [5:0] storageBank_0;\n'
              'logic [5:0] storageBank_1;\n'
              'logic [5:0] storageBank_4;\n'
              'logic [5:0] storageBank_3;\n'
              'logic [5:0] storageBank_2;\n'
              'logic [5:0] storageBank_5;\n'
              'logic [5:0] storageBank_6;\n'
              '//  combinational\n'
              'always_comb begin\n'
              '  if((~rd_en_0)) begin\n'
              '      rd_data_0 = 6\'h0;\n'
              '  end   else begin\n'
              '      case (rd_addr_0) \n'
              '          3\'h0 : begin\n'
              '              rd_data_0 = storageBank_0;\n'
              '          end\n'
              '          3\'h1 : begin\n'
              '              rd_data_0 = storageBank_1;\n'
              '          end\n'
              '          3\'h2 : begin\n'
              '              rd_data_0 = storageBank_2;\n'
              '          end\n'
              '          3\'h3 : begin\n'
              '              rd_data_0 = storageBank_3;\n'
              '          end\n'
              '          3\'h4 : begin\n'
              '              rd_data_0 = storageBank_4;\n'
              '          end\n'
              '          3\'h5 : begin\n'
              '              rd_data_0 = storageBank_5;\n'
              '          end\n'
              '          3\'h6 : begin\n'
              '              rd_data_0 = storageBank_6;\n'
              '          end\n'
              '          default : begin\n'
              '              rd_data_0 = 6\'h0;\n'
              '          end\n'
              '      endcase\n'
              '\n'
              '  end \n'
              '\n'
              'end\n'
              '\n'
              '//  sequential\n'
              'always_ff @(posedge clk) begin\n'
              '  if(reset) begin\n'
              '      storageBank_0 <= 6\'h0;\n'
              '      storageBank_1 <= 6\'h0;\n'
              '      storageBank_2 <= 6\'h0;\n'
              '      storageBank_3 <= 6\'h0;\n'
              '      storageBank_4 <= 6\'h0;\n'
              '      storageBank_5 <= 6\'h0;\n'
              '      storageBank_6 <= 6\'h0;\n'
              '  end   else begin\n'
              '      if((wr_en_0 & (wr_addr_0 == 3\'h0))) begin\n'
              '          storageBank_0 <= wr_data_0;\n'
              '      end \n'
              '\n'
              '      if((wr_en_0 & (wr_addr_0 == 3\'h1))) begin\n'
              '          storageBank_1 <= wr_data_0;\n'
              '      end \n'
              '\n'
              '      if((wr_en_0 & (wr_addr_0 == 3\'h2))) begin\n'
              '          storageBank_2 <= wr_data_0;\n'
              '      end \n'
              '\n'
              '      if((wr_en_0 & (wr_addr_0 == 3\'h3))) begin\n'
              '          storageBank_3 <= wr_data_0;\n'
              '      end \n'
              '\n'
              '      if((wr_en_0 & (wr_addr_0 == 3\'h4))) begin\n'
              '          storageBank_4 <= wr_data_0;\n'
              '      end \n'
              '\n'
              '      if((wr_en_0 & (wr_addr_0 == 3\'h5))) begin\n'
              '          storageBank_5 <= wr_data_0;\n'
              '      end \n'
              '\n'
              '      if((wr_en_0 & (wr_addr_0 == 3\'h6))) begin\n'
              '          storageBank_6 <= wr_data_0;\n'
              '      end \n'
              '\n'
              '  end \n'
              '\n'
              'end\n'
              '\n'
              'endmodule : RegisterFile\n'
              '\n'
              '////////////////////\n'
              '\n'
              'module Fifo(\n'
              'input logic readEnable,\n'
              'input logic [5:0] writeData,\n'
              'input logic reset,\n'
              'input logic clk,\n'
              'input logic writeEnable,\n'
              'output logic [5:0] readData,\n'
              'output logic error,\n'
              'output logic empty,\n'
              'output logic [2:0] occupancy,\n'
              'output logic full\n'
              ');\n'
              'logic [2:0] rd_addr_0;\n'
              'logic [2:0] wr_addr_0;\n'
              'logic wr_en_0;\n'
              'logic rd_en_0;\n'
              'logic [5:0] data;\n'
              'logic peekWriteData;\n'
              'logic matchedPointers;\n'
              'logic bypass;\n'
              'assign rd_en_0 = 1\'h1;\n'
              'assign readData = peekWriteData ? writeData : data;  // mux\n'
              'RegisterFile  rf(.wr_data_0(writeData),.rd_addr_0(rd_addr_0),.wr_addr_0(wr_addr_0),.clk(clk),.reset(reset),.wr_en_0(wr_en_0),.rd_en_0(rd_en_0),.rd_data_0(data));\n'
              'assign matchedPointers = wr_addr_0 == rd_addr_0;  // equals\n'
              'assign empty = matchedPointers & (~full);  // and_\n'
              '//  sequential\n'
              'always_ff @(posedge clk) begin\n'
              '  if(reset) begin\n'
              '      wr_addr_0 <= 3\'h0;\n'
              '      rd_addr_0 <= 3\'h0;\n'
              '      full <= 1\'h0;\n'
              '  end   else begin\n'
              '      if((~bypass)) begin\n'
              '          wr_addr_0 <= (writeEnable ? ((wr_addr_0 == 3\'h6) ? 3\'h0 : (wr_addr_0 + 3\'h1)) : wr_addr_0);\n'
              '          rd_addr_0 <= (readEnable ? ((rd_addr_0 == 3\'h6) ? 3\'h0 : (rd_addr_0 + 3\'h1)) : rd_addr_0);\n'
              '      end \n'
              '\n'
              '      full <= ((full & (writeEnable == readEnable)) | (((rd_addr_0 == ((wr_addr_0 == 3\'h6) ? 3\'h0 : (wr_addr_0 + 3\'h1))) & writeEnable) & 
(~readEnable)));\n'
              '  end \n'
              '\n'
              'end\n'
              '\n'
              'assign wr_en_0 = writeEnable & (~bypass);  // and__3\n'
              'assign bypass = (empty & readEnable) & writeEnable;  // and__4\n'
              'assign peekWriteData = empty & writeEnable;  // and__6\n'
              'assign error = ((empty & readEnable) & (~writeEnable)) | ((full & writeEnable) & (~readEnable));  // or__0\n'
              '//  sequential_0\n'
              'always_ff @(posedge clk) begin\n'
              '  if(reset) begin\n'
              '      occupancy <= 3\'h0;\n'
              '  end   else begin\n'
              '      unique case (({writeEnable,readEnable})) \n'
              '          2\'h2 : begin\n'
              '              occupancy <= (occupancy + 3\'h1);\n'
              '          end\n'
              '          2\'h1 : begin\n'
              '              occupancy <= (occupancy - 3\'h1);\n'
              '          end\n'
              '          default : begin\n'
              '              occupancy <= occupancy;\n'
              '          end\n'
              '      endcase\n'
              '\n'
              '  end \n'
              '\n'
              'end\n'
              '\n'
              'endmodule : Fifo'
     Which: does not contain '(wrPointer == 3\'h6'

  package:matcher                    expect
  test\configurator_test.dart 168:7  main.<fn>.<fn>
```

**Comment by mkorbel1** [comment]
@quekyj I dont see the test failing when I check out your branch locally?  Try pulling `main` in and rerun?

**Review by mkorbel1 — APPROVED** [review]
Looks good!

**Pull request merged**